DirectProgramming/DPC++FPGA/ReferenceDesigns/qri/src/qri.hpp

```diff
@@ -0,0 +1,155 @@
+#include <sycl/ext/intel/ac_types/ac_complex.hpp>
+#include <sycl/ext/intel/ac_types/ac_int.hpp>
+
+#include <chrono>
+#include <cstring>
+#include <type_traits>
+#include <vector>
+
+#include "memory_transfers.hpp"
+#include "streaming_qrd.hpp"
+#include "streaming_qri.hpp"
+#include "utils.hpp"
+
+// Forward declare the kernel names
+// (This prevents unwanted name mangling in the optimization report.)
+class QRIDDRToLocalMem;
+class QRD;
+class QRI;
+class QRILocalMemToDDRQ;
+
+template <unsigned columns,         // Number of columns in the input matrix
+          unsigned rows,            // Number of rows in the input matrix
+          unsigned raw_latency_qrd, // RAW latency for triangular loop
+                                    // optimization in the QRD kernel
+          unsigned raw_latency_qri, // RAW latency for triangular loop
+                                    // optimization in the QRI kernel
+          bool is_complex,          // Selects between ac_complex<T> and T
+                                    // datatype
+          typename T,               // The datatype for the computation
+          typename TT = std::conditional_t<is_complex, ac_complex<T>, T>
+                                    // TT will be ac_complex<T> or T depending
+                                    // on is_complex
+          >
+void QRIImpl(
+    std::vector<TT> &a_matrix,       // Input matrix to inverse
+    std::vector<TT> &inverse_matrix, // Output inverse matrix
+    sycl::queue &q,                  // Device queue
+    size_t matrices,                 // Number of matrices to process
+    size_t reps                      // Number of repetitions (for performance
+                                     // evaluation)
+) {
+
+  // Functional limitations
+  static_assert(std::is_same_v<T, float>, "only float datatype is supported");
+  static_assert(
+      rows >= columns,
+      "only rectangular matrices with rows>=columns are matrices supported");
+  static_assert((columns <= 512) && (columns >= 4),
+                "only matrices of size 4x4 to 512x512 are supported");
+
+  constexpr int kAMatrixSize = rows * columns;
+  constexpr int kInverseMatrixSize = rows * columns;
+  constexpr int kNumElementsPerDDRBurst = is_complex ? 4 : 8;
+
+  // Number of buffers to allocate to be able to read/compute/store without
+  // overlap.
+  constexpr short kNumBuffers = 3;
+
+  using PipeType = PipeTable<kNumElementsPerDDRBurst, TT>;
+
+  using a_matrix_pipe = sycl::ext::intel::pipe<class APipe, PipeType, 3>;
+  using q_matrix_pipe = sycl::ext::intel::pipe<class QPipe, PipeType, 3>;
+  using r_matrix_pipe = sycl::ext::intel::pipe<class RPipe, PipeType, 3>;
+  using inverse_matrix_pipe = sycl::ext::intel::pipe<class IPipe, PipeType, 3>;
+
+  // We will process 'kMatricesPerIter' number of matrices in each run of the
+  // kernel
+#if defined(FPGA_EMULATOR)
+  constexpr int kMatricesPerIter = 1;
+  assert(matrices % kMatricesPerIter == 0);
```

Comment:
This line is the same regardless of FPGA_EMULATOR being defined. Delete the two instances and move it out of the ifdef ?